movr.py

```diff
@@ -113,256 +123,205 @@ def add_vehicle_helper(session, city, owner_id, current_location, type, vehicle_
             session.add(pc)
             return pc.code
 
         return run_transaction(sessionmaker(bind=self.engine),
                                lambda session: add_promo_code_helper(session, code, description, expiration_time, rules))
 
-
     def apply_promo_code(self, user_city, user_id, promo_code):
+
         def apply_promo_code_helper(session, user_city, user_id, code):
             pc = session.query(PromoCode).filter_by(code=code).one_or_none()
             if pc:
                 # see if it has already been applied
-                upc = session.query(UserPromoCode).\
-                    filter_by(city = user_city, user_id = user_id, code = code).one_or_none() if self.multi_region else session.query(UserPromoCode).\
-                    filter_by(user_id = user_id, code = code).one_or_none()
+                upc = session.query(UserPromoCode).filter_by(city=user_city, user_id=user_id, code=code).one_or_none(
+                ) if self.multi_region else session.query(UserPromoCode).filter_by(user_id=user_id, code=code).one_or_none()
                 if not upc:
-                    upc = UserPromoCode(city = user_city, user_id = user_id, code = code)
+                    upc = UserPromoCode(
+                        city=user_city, user_id=user_id, code=code)
                     session.add(upc)
 
         run_transaction(sessionmaker(bind=self.engine),
-                               lambda session: apply_promo_code_helper(session, user_city, user_id, promo_code))
-
-    def multi_query_helper(session, queries):
-        for query in queries:
-            session.execute(query)
+                        lambda session: apply_promo_code_helper(session, user_city, user_id, promo_code))
 
     def run_queries_in_separate_transactions(self, queries):
         for query in queries:
-            run_transaction(sessionmaker(bind=self.engine),
-                            lambda session: session.execute(query))
+            try:
+                run_transaction(sessionmaker(bind=self.engine),
+                                lambda session: session.execute(query))
+            except ProgrammingError as err:
+                if 'Duplicate' in str(err):
+                    logging.info(
+                        "The following query will be skipped, due to a duplicate object error:")
+                    logging.info(str(query))
+                    continue
+                else:
+                    raise err
 
     ##############
     # MULTI REGION TRANSFORMATIONS
     ################
 
+    def assign_regions(self, regions):
+        DEFAULT_REGION_MAP = {
+            'us_east': ['new york', 'boston', 'washington dc'],
+            'us_west': ['san francisco', 'seattle', 'los angeles'],
+            'us_central': ['chicago', 'detroit', 'minneapolis'],
+            'eu_west': ['amsterdam', 'paris', 'rome']
+        }
+        region_map = {}
+
+        for region in regions:
+            if 'east' in region:
```

Comment:
there are other easts in regions: ap-east, eu-east, etc - we need to be more specific here

Reply:
same for central. In the current movr, users can specify the region/city pairing, any reason we moved away from that pattern?

Reply:
I've updated this logic quite a bit to make regions, cities, and region-city pairs more dynamic.

Only regions specified as node localities (i.e., regions returned by `SHOW REGIONS`) are allowed.